File: job-search-agents/agents/demand_generation.py
"""
Agent 2: Demand Generation
Reads Agent 1 outputs + target_roles.md.
Outputs: content calendar, LinkedIn posts, outreach scripts, target companies.
"""

from .base import (
    get_client,
    parse_sections,
    read_input,
    read_output,
    stream_and_collect,
    write_output,
)

_SYSTEM = """You are a B2B content strategist and outbound growth expert who has helped
hundreds of professionals build personal brands that generate inbound job opportunities.
You specialise in:

- Creating high-signal, platform-native content that builds authority (not noise)
- Turning lived experience into insights, stories, and POVs that resonate with decision-makers
- Designing warm outreach that feels like a genuine connection, not a cold pitch
- Mapping decision-makers at target companies and identifying the right entry points
- Engineering visibility loops that compound over 30-90 days

Your outputs are immediately actionable — copy-paste ready, not templates."""

_OUTPUT_INSTRUCTIONS = """
Using the positioning and career data above, produce all sections below.

===BEGIN 02_content_calendar.md===
## 4-Week LinkedIn Content Calendar

For each week provide a table:

| Day | Post Type | Topic/Hook | Goal |
|-----|-----------|------------|------|

Post types: Story, Insight, POV, Tactical How-To, Engagement Question, Case Study
Goals: Authority, Visibility, Network Growth, Recruiter Signal, Warm Leads

Include a brief weekly theme and 1-sentence strategic rationale per week.
===END 02_content_calendar.md===

===BEGIN 02_posts.md===
Write 12 ready-to-publish LinkedIn posts. Mix of types. Each post:

---
**Post N — [Type]: [Hook]**

[Full post text, 150-300 words, formatted with line breaks for readability.
No hashtag spam — max 3 targeted hashtags at end.]

*Strategic note: [1 sentence on why this post builds their positioning]*

---
===END 02_posts.md===

===BEGIN 02_outreach_scripts.md===
## Outreach Playbook

### Template 1: Recruiter Inbound (LinkedIn InMail)
[Subject line + 100-word message]

### Template 2: Warm Network Reconnect
[Subject line + 150-word message — someone they know but haven't spoken to in 2+ years]

### Template 3: Decision-Maker Cold Outreach (LinkedIn)
[Subject line + 120-word message — hiring manager or team lead, not HR]

### Template 4: Referral Ask
[Email + 100-word message — asking a mutual contact for an intro]

### Template 5: Post-Application Follow-Up
[Email + 80-word message — 5 days after applying with no response]

### Follow-Up Sequence (for Templates 2 & 3)
- Follow-up 1 (Day 5): [50 words]
- Follow-up 2 (Day 12): [40 words — final nudge, leaves door open]
===END 02_outreach_scripts.md===

===BEGIN 02_target_companies.md===
## Target Company List

Based on the target roles and positioning, identify 15-20 specific companies.
For each:

| Company | Why It's a Fit | Likely Role Title | Entry Point | Priority |
|---------|---------------|-------------------|-------------|----------|

After the table, write a **Contact-Finding Strategy** section:
1. How to find the right hiring manager on LinkedIn (search strings)
2. How to identify warm paths in (mutual connections, alumni, previous colleagues)
3. How to time outreach around company signals (funding, launches, job postings)
===END 02_target_companies.md===
"""


def run() -> None:
    print("\n=== Agent 2: Demand Generation ===")
    client = get_client()

    print("Reading inputs...")
    target_roles = read_input("target_roles.md")
    narrative = read_output("01_master_narrative.md")
    angles = read_output("01_positioning_angles.md")
    keyword_map = read_output("01_keyword_map.md")

    user_content = f"""## MASTER NARRATIVE (from Agent 1)
{narrative}

## POSITIONING ANGLES (from Agent 1)
{angles}

## KEYWORD MAP (from Agent 1)
{keyword_map}

## TARGET ROLES & INDUSTRIES
{target_roles}

---
{_OUTPUT_INSTRUCTIONS}"""

    print("Calling Claude (30-90 seconds)...\n")
    response = stream_and_collect(client, _SYSTEM, user_content)

    print("\nParsing outputs...")
    sections = parse_sections(response)
    expected = [
        "02_content_calendar.md",
        "02_posts.md",
        "02_outreach_scripts.md",
        "02_target_companies.md",
    ]
    for filename in expected:
        if filename in sections:
            write_output(filename, sections[filename])
        else:
            print(f"  WARNING: Could not parse section for {filename}")

    print("\nAgent 2 complete — 4 files written to outputs/")

File: src/agents/prompts/demandGeneration.ts
export const SYSTEM_PROMPT = `You are a B2B demand generation strategist who specializes in personal brand building for job seekers. Your approach: create visibility and credibility so opportunities come inbound instead of relying on cold applications.

You write LinkedIn content that sounds like a smart practitioner thinking out loud — not a motivational poster or a job seeker desperately broadcasting their search. The best posts are specific, opinionated, and teach something in 60 seconds.

Outreach messages you write are warm, high-context, and one-directional — they open conversations, they don't pitch.

Format your output with clear labeled sections using markdown-style headers (## Section Name). Use bullet points for lists. Include the actual post text verbatim in code blocks.`

export interface DemandGenerationInputs {
  narrative: string
  targetCompanies: string
  expertiseTopics: string
  contentTone: string
  targetAudience: string
}

export function buildUserMessage(inputs: DemandGenerationInputs): string {
  return `Here is my positioning context. Build a full demand generation package.

## MY NARRATIVE / POSITIONING
${inputs.narrative}

## TARGET COMPANIES
${inputs.targetCompanies}

## EXPERTISE TOPICS I CAN SPEAK TO
${inputs.expertiseTopics}

## CONTENT TONE PREFERENCE
${inputs.contentTone}

## TARGET AUDIENCE (who I want to reach)
${inputs.targetAudience}

---

Produce the following, clearly labeled:

## 1. Content Strategy (1-paragraph brief)
What angle I should own on LinkedIn. Why it works for reaching hiring managers in this space.

## 2. Weekly Content Calendar
A 5-post calendar for one week. For each post: Day · Format · Topic · Hook (first line). Table format.

## 3. Five LinkedIn Posts (full text)
Write all 5 posts in full. Each post should be under 250 words. Lead with a strong hook. End with a question or call to action. Format each in a code block.

## 4. Comment Bank (5–8 comments)
Short, smart comments I can leave on posts from people at target companies. Each one adds genuine value or a sharp observation — not flattery.

## 5. Outreach DM Scripts (3 variants)
Three different outreach messages for different contexts:
- Variant A: Reaching out to a hiring manager I haven't met
- Variant B: Following up after engaging with someone's content
- Variant C: Reconnecting with a dormant contact

Each should be under 80 words. Warm, not salesy. Opens a door, doesn't ask for a job.`
}
